subprojects/core/src/main/java/org/gradle/api/internal/changedetection/rules/TaskUpToDateState.java

```diff
@@ -70,7 +75,7 @@ public TaskStateChanges getInputFilesChanges() {
      * Returns if any output files have been changed, added or removed.
      */
     public boolean hasAnyOutputFileChanges() {
-        return outputFileChanges.hasAnyChanges();
+        return noHistoryState.iterator().hasNext() || outputFilePropertyNameChanges.iterator().hasNext() || outputFileChanges.hasAnyChanges();
     }
```

Comment:
Why do we need to check for no history here now? If we have to do that, I think we should do it via an explicit method (say, `noHistoryState.hasHistory()` or something), or perhaps just check if `lastExecution == null`.

Reply:
Before my change, output file changes was yielding descriptive changes for
- no previous execution
- output file property changes
- changed output files

This is now tracked by different states.

Reply:
I see... This might be a little easier to follow I think:

```java
    public boolean hasAnyOutputFileChanges() {
        if (lastExecution == null) {
             return true;
        }
        return outputFilePropertyNameChanges.iterator().hasNext()
            || outputFileChanges.hasAnyChanges();
    }
```

Reply:
Should be clearer now.